src/Algebra/Properties/CancellativeCommutativeSemiring.agda

```diff
@@ -0,0 +1,72 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Some theory for CancellativeCommutativeSemiring.
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+open import Algebra using (CancellativeCommutativeSemiring)
+import Algebra.GCD
+import Algebra.Primality
+import Algebra.Properties.Semigroup.Divisibility as SemigroupDiv
+import Algebra.Properties.Semiring.Divisibility as SemiringDiv
+open import Data.Product using (_×_; _,_; proj₁; proj₂)
+open import Data.Sum using (_⊎_; inj₁; inj₂)
+open import Function using (case_of_; flip)
+open import Level using (_⊔_)
+open import Relation.Binary using (Rel; Symmetric; Setoid)
+open import Relation.Unary using (Pred)
+
+module Algebra.Properties.CancellativeCommutativeSemiring
+  {α ℓ} (R : CancellativeCommutativeSemiring α ℓ)
+  (open CancellativeCommutativeSemiring R renaming (Carrier to A))
+  where
+
+open SemigroupDiv *-semigroup using (_∣_; _∤_; ∣-trans)
+open SemiringDiv semiring using (∣nonzero⇒≉0)
+open Algebra.Primality _≈_ _*_ 0# 1# using (IsIrreducible; Coprime)
+open Algebra.GCD _≈_ _*_ using (GCD)
+
+0∤1 : 0# ≉ 1# → 0# ∤ 1#
+0∤1 0≉1 (q , q*0≈1) = 0≉1 (trans (sym (zeroʳ q)) q*0≈1)
+
+------------------------------------------------------------------------------
+-- Properties of Irreducibilty, Coprime.
+
+irreducible≉0 : 0# ≉ 1# → (∀ {p} → IsIrreducible p → p ≉ 0#)
+irreducible≉0 0≉1 (_ , chooseInvertible) p≈0 =
+  let p≈0*0 = trans p≈0 (sym (zeroˡ 0#)) in
+  case chooseInvertible p≈0*0 of λ
+  { (inj₁ 0∣1) → 0∤1 0≉1 0∣1
+  ; (inj₂ 0∣1) → 0∤1 0≉1 0∣1
+  }
+
+coprimeWithInvertible : ∀ {x} y → x ∣ 1# → Coprime x y
```

Comment:
Because both of the notions ``Invertible x`` and ``x ∣ 1#``
(``x ∣ ε`` of monoid) are defined in algebra as ``∃ \q → q * x ≈ 1#``.  
(``∃ \q → q ∙ x ≈ ε``).  

"coprimeToInvertible" is shorter, but will not do, because the reader could think of "To" as of conversion to something.

Reply:
What is this notion of `Invertible`? I'm not aware of anything called that in the library?

Reply:
Textbooks write like this:
An element `x` in a Monoid is called left-invertible iff
there exists q such that  ``q * x ≈ ε``.

And this coincides with our definitions of  ``_∣_,  x ∣ ε``.

Firther, in CommutativeMonoid, and in any Group,
left-invertible <==> right-invertible,
but not in the general case.

An element  x  in a Monoid  is called invertible iff
```
∃ \q → q ∙ x ≈ ε  ×  x ∙ q ≈ ε
```
, that is x has a left inverse q which is also a right inverse for x.

Now I think:
left-invertibe x  is expressed as  ``x ∣ ε``
- which is "x divides ε from right"
``(∃ \q → q ∙ x ≈ ε)``.  

So, (right-invertibe x)  needs to correspond to "x divides ε from left".

Assuming all this, I introduce _divisibility from left_   ``_∣ˡ_`` 
and divisibilty from right ``_∣ʳ_``.
They are equivalent in any commutative magma and in any group, but not in the general case.
I put
```
x ∣ˡ y = ∃ λ q → (x ∙ q) ≈ y    -- x divides y from left                          
x ∣ʳ y = ∃ λ q → (q ∙ x) ≈ y    -- x divides y from right                         

_∣_ = _∣ʳ_
```.
The latter is set because in many domains left and right divisibilities
are equivalent (as in ℕ and ℤ), and there it is simpler to write `_∣_`.
See Algebra.Divisibility.
Is this all right?